book/accessibility.md

```diff
@@ -770,280 +849,327 @@ class Browser:
+class Tab:
+    def load(self, url, body=None):
+        self.focus = None
+        # ...
 ```
 
-Each time `tab` is pressed, the browser should advance focus to the next thing
-in order. This will first require a definition of which elements are
-focusable:
+We now have the ability to focus on links, buttons, and text entries.
+But as with any browser feature, it's worth asking whether web page
+authors should be able to customize it. With keyboard navigation, the
+author might want certain links not to be focusable (like "permalinks"
+to a section heading, which would just be noise to most users), or
+might want to change the order in which the user tabs through
+focusable items.
 
-``` {.python expected=False}
-def is_focusable(node):
-    return node.tag == "input" or node.tag == "button" \
-    	or node.tag == "a"
+Browsers support the `tabindex` HTML attribute to make this possible.
+The `tabindex` attribute is a number; elements with a negative value
+aren't focusable and smaller numbers come before larger numbers and
+elements without a `tabindex` attribute. To implement that, we need to
+sort the focusable elements by tab index, so we need a function that
+returns the tab index:
+
+``` {.python}
+def get_tabindex(node):
+    return int(node.attributes.get("tabindex", "9999999"))
 ```
 
-And then each iterating through them:
+The default value, "9999999", makes sure that elements without a
+`tabindex` attribute sort after ones with the attribute. Now we can
+sort by `get_tabindex` in `advance_tab`:
 
 ``` {.python}
 class Tab:
     def advance_tab(self):
         focusable_nodes = [node
             for node in tree_to_list(self.nodes, [])
-            if isinstance(node, Element) and is_focusable(node)]
-        if not focusable_nodes:
-            self.apply_focus(None)
-        elif not self.focus:
-            self.apply_focus(focusable_nodes[0])
-        else:
-            i = focusable_nodes.index(self.focus)
-            if i < len(focusable_nodes) - 1:
-                self.apply_focus(focusable_nodes[i+1])
-        self.set_needs_render()
+            if isinstance(node, Element) and is_focusable(node)
+            and get_tabindex(node) >= 0]
+        focusable_nodes.sort(key=get_tabindex)
+        # ...
+```
+
+Since Python's sort is "stable", two elements with the same `tabindex`
+won't change their relative position in `focusable_nodes`.
+
+Additionally, elements with `tabindex` are automatically focusable,
+even if they aren't a link or a button or a text entry. That's useful,
+because that element might listen to the `click` event. To support
+this let's first extend `is_focusable` to consider `tabindex`:
+
+``` {.python}
+def is_focusable(node):
+    return node.tag in ["input", "button", "a"] \
+        or "tabindex" in node.attributes
 ```
 
- When `tab` is pressed and we're at the
-end of the focusable list, focus the address bar:
+Next, we need to make sure to send a `click` event when an element is
+activated:
```

Comment:
Add a sentence about the JS behavior, since just like click  enter can be preventDefaulted